flojoy/flojoy_node_venv.py

```diff
@@ -83,83 +82,99 @@ def _get_venv_executable_path(venv_path: os.PathLike | str) -> os.PathLike | str
+                    buffer_logs=True,
+                )
             )
-        )
-        proc = subprocess.Popen(
-            command,
-            stdout=logpipe_stdout.get_pipe_writer(),
-            stderr=logpipe_stderr.get_pipe_writer(),
-        )
-        # Poll the process until it finishes, while occasionally checking if there was a failure
-        # in the global cancel event
-        while proc.poll() is None:
-            if PipInstallThread._cancel_all_threads.is_set():
-                proc.terminate()
-                logger.error(
-                    f"Another thread has failed its pip install step, terminating pip install process for {venv_path}..."
+            logpipe_stdout = stack.enter_context(
+                LogPipe(
+                    logger,
+                    log_level=logging.INFO,
+                    mode=LogPipeMode.SUBPROCESS,
+                    buffer_logs=True,
                 )
-                break
-            sleep(0.1)
+            )
+            proc = subprocess.Popen(
+                command,
+                stdout=logpipe_stdout.get_pipe_writer(),
+                stderr=logpipe_stderr.get_pipe_writer(),
+            )
+            # Poll the process until it finishes, while occasionally checking if there was a failure
+            # in the global cancel event
+            while proc.poll() is None:
+                if PipInstallThread._cancel_all_threads.is_set():
+                    proc.terminate()
+                    logger.error(
+                        f"Another thread has failed its pip install step, terminating pip install process for {venv_path}..."
+                    )
+                    break
+                sleep(0.1)
 
-    venv_complete_path = os.path.realpath(os.path.join(venv_path, ".venv_is_complete"))
 
-    # The process was terminated due to the _cancel_all_threads event
-    if proc.returncode is None:
-        return
+        # The process was terminated due to the _cancel_all_threads event
+        if proc.returncode is None:
+            return
 
-    if proc.returncode != 0:
-        if os.path.exists(venv_complete_path):
-            os.remove(venv_complete_path)
-        shutil.rmtree(venv_path, ignore_errors=True)
-        logger.error(
-            f"Failed to install pip dependencies into virtual environment from the provided list: {pip_dependencies}. The virtual environment under {venv_path} has been deleted."
-        )
-        raise subprocess.CalledProcessError(
-            proc.returncode,
-            command,
-            output=logpipe_stdout.log_buffer.getvalue().encode("utf-8"),
-            stderr=logpipe_stdout.log_buffer.getvalue().encode("utf-8"),
-        )
+        if proc.returncode != 0:
+            # First wipe the .venv_is_complete file.
+            if os.path.exists(venv_is_complete_path):
+                os.remove(venv_is_complete_path)
+            # Then the rest
+            shutil.rmtree(venv_path, ignore_errors=True)
```

Comment:
Why can't we just wipe the virtual environment directly?
```suggestion
            # Wipe the virtual environment.
            shutil.rmtree(venv_path, ignore_errors=True)
```

Reply:
`rmtree` wipes directories recursively starting from leaves, and there exists an edge where the program exits before `rmtree` has finished, leaving the environment directory incomplete while `.venv_is_complete` is still present. I start by deleting that first, which renders the directory incomplete immediately

Reply:
Thanks, this is perfectly clear now. Then I think we can complete your comment with bits of your explanation.
```suggestion
            # First wipe the .venv_is_complete file to mark the directory as invalid
            # in case the deletion of the entire directory fails.
            if os.path.exists(venv_is_complete_path):
                os.remove(venv_is_complete_path)
            # Then delete the entire directory.
            shutil.rmtree(venv_path, ignore_errors=True)
```

What do you think?

Reply:
Agreed, thank you